src/replication.c

```diff
@@ -1214,7 +1189,34 @@ int replicationAppend(struct raft *r,
         goto err_after_acquire_entries;
     }
 
+    /* Possibly apply configuration changes as uncommitted. */
+    for (j = i; j < args->n_entries; j++) {
+        struct raft_entry *entry = &args->entries[j];
+        raft_index index = args->prev_log_index + 1 + j;
+        raft_term local_term = logTermOf(r->log, index);
+
+        assert(local_term != 0 && local_term == entry->term);
+
+        if (entry->type == RAFT_CHANGE) {
+            rv = membershipUncommittedChange(r, index, entry);
+            if (rv != 0) {
+                tracef("membershipUncommittedChange");
+                goto err_after_acquire_entries;
+            }
+        }
+    }
```

Comment:
I think it would be a tad simpler and more explicit to call `membershipUncommittedChange()` in the for loop above, right after the `logAppend()` call [here](https://github.com/cole-miller/raft/blob/a8eb49061d2775b56ba8defd12d7d64fcfa2832c/src/replication.c#L1165).

That will save a bit of code (since the two loops should be the same) and make it more clear that configuration changes are applied immediately upon appending them to the in-memory log.

Reply:
I didn't look much into what type of failures could happen into this specific spot, but as a general rule I'd say we have 2 types of failures in raft:

- Failures that make it totally impossible to proceed (for example failing to allocate memory that we absolutely need): in this case we should basically stop-the-line and transition to `RAFT_UNAVAILABLE` (user code will typically abort the process).

- Failures that can be retried (that's typically I/O): in this case we should have some mechanism to retry (for example if the disk is currently full, but eventually the operator frees some of it). We currently handle that fine for network I/O, but not that much for disk I/O, and it's an area that should be improved.

In both cases, I don't think there's any need to restore the old (current, committed) configuration: we should continue operating with the uncommitted one and deal with I/O errors by retrying.

Note though, that I'm not totally convinced anymore that it's a good idea to immediately apply the new configuration at all, before persisting it. See the example scenario [here](https://github.com/canonical/raft/pull/465#issuecomment-1667960666) and the follow-up notes.